text/0031-integ-testing.md

```diff
@@ -0,0 +1,483 @@
+
+### `cdk-deploy`
+
+This jsii module provides the ability to programmatically deploy and destroy AWS CDK apps.
+
+A CDK app can be deployed by calling -
+
+```ts
+import { App } from 'cdk-deploy';
+
+const app = new App('/path/to/cdk/app');
+app.deploy();
+...
+app.destroy();
+```
+
+The path provided here can be a path to a project that contains a `cdk.json` file or a path
+to a CDK app [cloud assembly].
+
+The `deploy()` and `destroy()` APIs will block further program execution until the deployment
+or destruction complete. If the action fails, the API will throw an error.
+Asynchronous support for `deploy()` and `destroy()` are not available yet.
+
+> TBD: configurable options of deploy and destroy
+
+By default, the APIs print their output and progress to stdout. This can be configured
+using options on the APIs.
+
+```ts
+app.deploy({
+  stdout: false,
+  outfile: '/path/to/file' // path to file where the progress should be printed
+});
+```
+
+#### Credentials
+
+This module uses the AWS SDK for Javascript under the hood to communicate with CloudFormation.
+We support most of the common ways in which [this SDK allows setting
+credentials](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/setting-credentials-node.html).
+
+Credentials are loaded from the following and in the given order of precedence,
+
+1. [Environment variables](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/loading-node-credentials-environment.html)
+1. [Shared credentials file](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/loading-node-credentials-shared.html)
+1. Execution platform credentials provider, such as,
+   [IAM roles for Amazon EC2](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/loading-node-credentials-iam.html),
+   [IAM roles for AWS Lambda](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/loading-node-credentials-lambda.html),
+   or Amazon ECS credentials provider.
+
+---
+
+Ticking the box below indicates that the public API of this RFC has been
+signed-off by the API bar raiser (the `api-approved` label was applied to the
+RFC pull request):
+
+```
+[ ] Signed-off by API Bar Raiser @xxxxx
+```
+
+## Public FAQ
+
+### What are we launching today?
+
+We are launch a new AWS CDK module, named `apptest`, that enables customers to write integration
+tests as part of their CDK app, and execute them on their developer machines or on a CI server
+as part of their continuous deployment.
+
+These integration tests can be used to verify that their CDK app or construct library deploys
+successfully on AWS CloudFormation. Users can add additional assertions to verify the behaviour
+of their app, such as, invoking an API endpoint or sending a request to a load balancer.
+
+### Why should I use this feature?
+
+This feature is beneficial to users who own a CDK app and want to run assertions on the deployed
+app in their pre-production stages as part of a CI/CD pipeline.
```

Comment:
So these tests are not deployed as part of the main application? If not, then they cannot trigger a rollback if any of the tests fail, which seems like a nice feature.

> pre-production stages

Why not allow tests against production? Just curious :)

We are currently using step function and/or lambdas to run integration tests.  Since they are part of the stack, they can roll it back if anything fails and we can run it against any deployment stage.  Maybe this RFC isn't meant to address this use case?

This RFC is very interesting! It'll be helpful no matter what.  Just trying to grasp how it is run and when (EG: how would this run in CodePipeline?).

Reply:
> Why not allow tests against production? Just curious :)

You are right. Take a look at the latest revision. I've included this.

Reply:
Think I caught your updates, looks like it is possible to include these in your main stack in order to trigger rollbacks, thanks! 👍 

For when it isn't part of your main stack, what's the expected workflow in a pipeline like CodePipeline?

- CloudFormation Action to deploy the main application that we want to test
- CodeBuild that runs CDK deploy/destroy or Another CloudFormation Action to deploy... and maybe a CloudFormation Action to destroy? (or maybe just keep it? No real cost if all serverless tech)

Think my concern about CodeBuild is if the deployment isn't in the same account, we have to do some role assumes.

Reply:
> what's the expected workflow in a pipeline like CodePipeline?

You would run the tests as you would any other tests in your pipeline, which is probably CodeBuild.

A future addendum to this RFC will address this when [using CDK Pipelines](https://github.com/aws/aws-cdk/tree/master/packages/%40aws-cdk/pipelines#cdk-pipelines) and make this trivial. I've taken this out of scope to keep the RFC contained.

If you use your own CI/CD mechanism, you will need to set up cross-account assume role, yes.